### ISSUE TITLE
Decide whether / how to ban SAC + inplace ops in eager

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#119590]
SAC exists as an API today (see [code](https://github.com/pytorch/pytorch/blob/main/torch/utils/checkpoint.py#L1256)), but:

(1) it "context" fn has a pt2-specific name
(1) We have a warning in the docs that it should only be used with `torch.compile`
(2) We have no warning or error that gets emitted at runtime if you actually use SAC with eager mode.

My understanding is that the main issue with always-allowing SAC to be used in eager has to do with handling for inplace ops. More diagnosis was in this issue: https://github.com/pytorch/pytorch/issues/113737

I think it can be summarized by this repro, where eager mode vs. "eager mode + SAC" produce different outputs, when an inplace op is involved:
```
import torch
from torch._custom_op.functional import register_functional_op
import torch.utils.checkpoint
from torch.utils.checkpoint import checkpoint, _pt2_selective_checkpoint_context_fn_gen

def custom_policy(mode, func, *args, **kwargs):
    return func in [torch.ops.aten.mm.default]

def selective_checkpointing_context_fn():
    return _pt2_selective_checkpoint_context_fn_gen(custom_policy)

def gn(x, y):
    return torch.selu_(torch.matmul(x, y))

def fn(x, y):
    return torch.utils.checkpoint.checkpoint(
        gn,
        x,
        y,
        use_reentrant=False,
        context_fn=selective_checkpointing_context_fn,
    )

x = torch.arange(16, dtype=torch.float32, requires_grad=True).reshape(4, 4).detach().requires_grad_(True)
y = torch.arange(16, dtype=torch.float32, requires_grad=True).reshape(4, 4).detach().requires_grad_(True)

out1 = gn(x, y)
print(out1)
out1.sum().backward()
print(out1)

out2 = fn(x, y)
print(out2)
# With SAC + eager mode:
# (1) "out" is an activation saved for backward
# (2) selu_() is part of the recompute, which mutates out **again**, during the backward pass!
# Invoking the backward will mutate out!
out2.sum().backward()
print(out2)
# False
print(torch.allclose(out1, out2))
```

Just to collect some possible options:

(1) [easiest] Ban SAC completely in eager
(2) [medium] Ban SAC in eager whenever there are any inplace ops
(3) [hard?] figure out how to detect exactly the case when outputs/gradients would diverge without SAC, and ban those cases
(4) [hard?] figure out how to functionalize away an mutations in an SAC region that would have changed numerics.

cc @soulitzer @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @Lezcano @Varal7

**Comment by ezyang** [comment]
2 seems feasible especially with version counters

**Comment by soulitzer** [comment]
Related: https://github.com/pytorch/pytorch/issues/96319

**Comment by soulitzer** [comment]
Resolution: we used version counter to detect when the SAC cache is mutated between save and use.
